File: windows-native/DataMover.Core/Services/FanOutCopier.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;

namespace DataMover.Core.Services;

/// <summary>
/// [M1, 2026-09-06] Port 1:1 al FanOutCopier.swift (Mac) — vezi acolo
/// pentru comentariul complet de arhitectura/motiv. Pe scurt: inlocuieste
/// tiparul vechi din OffloadEngine.cs (`VerifyPair` re-citeste sursa PLUS
/// destinatia, per destinatie, dupa ce `CopyFileCancelable` a citit deja
/// o data sursa pentru copiere - 4 citiri ale sursei cu 2 destinatii) cu
/// UN SINGUR thread de citire din card, care distribuie fiecare bucata
/// catre N thread-uri de scriere prin `BlockingCollection&lt;T&gt;`
/// (colectia marginita nativa .NET - potrivire perfecta pentru ring
/// buffer cu backpressure, fara sa trebuiasca reinventata pe aceasta
/// platforma cum a fost nevoie pe Swift/AppKit).
/// </summary>
public sealed class FanOutError : Exception
{
    public FanOutError(string message) : base(message) { }
}

public sealed class FanOutDestResult
{
    public bool Success { get; init; }
    public string? Hash { get; init; }
    public long BytesWritten { get; init; }
    public Exception? Error { get; init; }
}

public sealed class FanOutResult
{
    public string SourceHash { get; init; } = "";
    public long BytesRead { get; init; }
    public Dictionary<string, FanOutDestResult> Destinations { get; init; } = new();
}

/// <summary>
/// Wrapper incremental unificat peste toate modelele de verificare —
/// extras din bucla existenta a lui `HashOfFile` (OffloadEngine.cs) ca sa
/// poata fi alimentat bucata-cu-bucata dintr-un flux extern (fan-out),
/// nu doar dintr-o bucla proprie care citeste un fisier de pe disc.
/// `HashOfFile` ramane neschimbat — foloseste in continuare propria bucla.
/// </summary>
public sealed class IncrementalHasher : IDisposable
{
    private readonly XxHash64? _xx;
    private readonly HashAlgorithm? _algo;
    private long _sizeOnlyCount;

    public IncrementalHasher(VerificationModel model)
    {
        if (model == VerificationModel.XxHash64)
        {
            _xx = new XxHash64();
        }
        else if (model != VerificationModel.SizeOnly)
        {
            _algo = model switch
            {
                VerificationModel.Md5 => MD5.Create(),
                VerificationModel.Sha1 => SHA1.Create(),
                VerificationModel.Sha256 => SHA256.Create(),
                VerificationModel.Sha512 => SHA512.Create(),
                _ => MD5.Create(),
            };
        }
    }

    public void Update(byte[] buffer, int count)
    {
        if (_xx != null) { _xx.Update(buffer, count); return; }
        if (_algo != null) { _algo.TransformBlock(buffer, 0, count, null, 0); return; }
        _sizeOnlyCount += count; // marime-doar: nu exista hash real, doar numaram
    }

    public string FinalizeHex()
    {
        if (_xx != null) return _xx.HexDigest();
        if (_algo != null)
        {
            _algo.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
            return Convert.ToHexString(_algo.Hash!).ToLowerInvariant();
        }
        return "";
    }

    public void Dispose() => _algo?.Dispose();
}

/// O bucata citita din sursa — `Length` poate fi mai mic decat
/// `Data.Length` la ultima bucata dintr-un fisier (buffer alocat per
/// bucata, nu reutilizat, ca fiecare consumator sa aiba propria referinta
/// stabila cat timp o proceseaza pe threadul lui).
public sealed record ChunkBuffer(byte[] Data, int Length);

public sealed class FanOutCopier
{
    private readonly string _sourcePath;
    private readonly IReadOnlyList<string> _destinationPaths;
    private readonly int _chunkSize;
    private readonly int _ringDepth;
    private readonly VerificationModel _model;
    private readonly CancelToken _cancel;
    private readonly PauseToken _pause;

    public FanOutCopier(string sourcePath, IReadOnlyList<string> destinationPaths, int chunkSize,
        VerificationModel model, CancelToken cancel, PauseToken pause, int ringDepth = 3)
    {
        _sourcePath = sourcePath;
        _destinationPaths = destinationPaths;
        _chunkSize = chunkSize;
        _ringDepth = Math.Max(2, ringDepth);
        _model = model;
        _cancel = cancel;
        _pause = pause;
    }

    public FanOutResult Run(Action<long> onBytesRead)
    {
        var queues = _destinationPaths.ToDictionary(
            d => d, d => new BlockingCollection<ChunkBuffer>(boundedCapacity: _ringDepth));
        var results = new ConcurrentDictionary<string, FanOutDestResult>();
        var failedDestinations = new HashSet<string>();

        var tasks = _destinationPaths.Select(dst => Task.Run(() =>
        {
            var queue = queues[dst];
            try
            {
                using var output = new FileStream(dst, FileMode.Create, FileAccess.Write, FileShare.None, _chunkSize);
                using var hasher = new IncrementalHasher(_model);
                long written = 0;
                foreach (var chunk in queue.GetConsumingEnumerable())
                {
                    output.Write(chunk.Data, 0, chunk.Length);
                    hasher.Update(chunk.Data, chunk.Length);
                    written += chunk.Length;
                }
                // [M2, 2026-09-06] Flush FIZIC obligatoriu inainte de a marca
                // fisierul OK - vezi comentariul complet din FanOutCopier.swift
                // (Mac) pentru motiv. `FileStream.Flush(true)` e documentat
                // oficial de Microsoft sa apeleze `FlushFileBuffers` la nivel
                // de OS pe Windows (nu doar bufferul intern .NET) - fara
                // P/Invoke necesar. O eroare aici opreste DOAR aceasta
                // destinatie, nu si celelalte.
                output.Flush(true);
                results[dst] = new FanOutDestResult { Success = true, Hash = hasher.FinalizeHex(), BytesWritten = written };
            }
            catch (Exception ex)
            {
                results[dst] = new FanOutDestResult { Success = false, Error = ex };
                try { queue.CompleteAdding(); } catch (ObjectDisposedException) { /* deja completat */ }
            }
        })).ToArray();

        // Thread-ul de CITIRE — singurul care atinge sursa. Ruleaza sincron
        // pe thread-ul apelant (deja de fundal in OffloadRunner).
        using var input = new FileStream(_sourcePath, FileMode.Open, FileAccess.Read, FileShare.Read,
            _chunkSize, FileOptions.SequentialScan);
        using var sourceHasher = new IncrementalHasher(_model);
        long bytesRead = 0;
        Exception? readError = null;
        try
        {
            while (true)
            {
                if (_cancel.IsCancelled) throw new OffloadCancelledException();
                while (_pause.IsPaused && !_cancel.IsCancelled) Thread.Sleep(50);
                if (_cancel.IsCancelled) throw new OffloadCancelledException();

                var buffer = new byte[_chunkSize];
                int read = input.Read(buffer, 0, _chunkSize);
                if (read == 0) break;

                sourceHasher.Update(buffer, read);
                bytesRead += read;
                onBytesRead(read);

                var chunk = new ChunkBuffer(buffer, read);
                foreach (var (dst, queue) in queues)
                {
                    if (failedDestinations.Contains(dst)) continue;
                    try { queue.Add(chunk); }
                    catch (InvalidOperationException)
                    {
                        // consumatorul acelei destinatii a esuat deja si a
                        // inchis coada (CompleteAdding) - ignoram tacit
                        // pentru restul transferului acestui fisier, dar
                        // NU oprim celelalte destinatii sanatoase.
                        failedDestinations.Add(dst);
                    }
                }
            }
        }
        catch (Exception ex)
        {
            readError = ex;
        }
        finally
        {
            foreach (var q in queues.Values)
            {
                try { q.CompleteAdding(); } catch (InvalidOperationException) { /* deja completat de esec */ }
            }
        }

        Task.WaitAll(tasks);

        if (readError != null) throw readError;

        return new FanOutResult
        {
            SourceHash = sourceHasher.FinalizeHex(),
            BytesRead = bytesRead,
            Destinations = results.ToDictionary(kv => kv.Key, kv => kv.Value),
        };
    }
}

File: mac-native/Sources/DataMoverMac/FanOutCopier.swift
import Foundation
import CryptoKit
import Darwin

/// [M1, 2026-09-06] Motor de copiere "citire unică, scriere multiplă" —
/// înlocuiește tiparul vechi (un `DestinationJob` separat per destinație,
/// fiecare citind sursa de la zero pentru copiere ȘI din nou pentru
/// verificare) cu UN SINGUR thread de citire din card, care distribuie
/// fiecare bucată către N thread-uri de scriere + calculează hash-ul
/// sursei O SINGURĂ DATĂ, din exact aceleași bucăți citite.
///
/// DE CE (găsit direct în cod, nu presupus): vechiul `processOne` apela
/// `copyFileCancelable` (1 citire completă a sursei) apoi `verifyPair`
/// (încă o citire COMPLETĂ a sursei pentru `srcHash`, plus o citire a
/// destinației pentru `dstHash`) — PER destinație. Cu 2 destinații, un
/// card se citea efectiv de 4 ori. Motorul de aici citește sursa O DATĂ,
/// indiferent de câte destinații există, și calculează hash-ul fiecărei
/// destinații ÎN TIMP CE SE SCRIE (nu printr-o recitire ulterioară de pe
/// disc) — elimină și cele N citiri ale destinațiilor.
///
/// ARHITECTURĂ: un `BoundedChunkQueue` per destinație (ring buffer marginit,
/// adâncime implicit 3 bucăți) — thread-ul de citire împinge aceeași
/// instanță `Data` (Copy-on-Write, niciun consumator n-o mutează, deci nu
/// se copiază efectiv de N ori în memorie) în toate cozile; câte un thread
/// de scriere per destinație consumă din propria coadă, scrie pe disc și
/// actualizează propriul hash incremental. Adâncimea marginită e
/// backpressure-ul cerut de Regula 21 („fără read-ahead nemărginit care ar
/// acumula date nescrise în RAM") — dacă o destinație e lentă (HDD extern),
/// coada ei se umple, thread-ul de citire se blochează la următorul
/// `push`, deci viteza reală de citire de pe card se aliniază la CEA MAI
/// LENTĂ destinație, exact comportamentul de siguranță cerut.
enum FanOutError: Error {
    case cannotOpenSource(String)
    case cannotOpenDestination(String)
}

/// Rezultatul per-destinație al unei copieri fan-out — fie succes cu
/// hash-ul calculat la scriere, fie eroarea specifică acelei destinații
/// (o destinație poate eșua — disc plin, deconectat — fără să oprească
/// scrierea către celelalte).
enum FanOutDestOutcome {
    case success(hash: String, bytesWritten: Int64)
    case failure(Error)
}

struct FanOutResult {
    /// Hash-ul sursei, calculat o singură dată, din bucățile citite.
    let sourceHash: String
    let bytesRead: Int64
    /// Cheie = calea destinației.
    let destinations: [String: FanOutDestOutcome]
}

/// Wrapper incremental unificat peste toate modelele de verificare —
/// extras din `genericHash`/`xxhash64OfFile` (OffloadEngine.swift) ca să
/// poată fi alimentat bucată-cu-bucată dintr-un flux extern, nu doar
/// dintr-o buclă proprie de citire a unui fișier de pe disc. Funcțiile
/// EXISTENTE rămân neschimbate — folosesc în continuare propriile bucle
/// (verificarea unui fișier deja existent la reluare, de exemplu, tot
/// citește fișierul o dată, nu are ce economisi).
enum IncrementalHasher {
    case xx(XXHash64)
    case md5(Insecure.MD5)
    case sha1(Insecure.SHA1)
    case sha256(SHA256)
    case sha512(SHA512)
    case sizeOnly(Int64)

    init(model: VerificationModel) {
        switch model {
        case .xxhash64: self = .xx(XXHash64())
        case .md5: self = .md5(Insecure.MD5())
        case .sha1: self = .sha1(Insecure.SHA1())
        case .sha256: self = .sha256(SHA256())
        case .sha512: self = .sha512(SHA512())
        case .sizeOnly: self = .sizeOnly(0)
        }
    }

    mutating func update(_ data: Data) {
        switch self {
        case .xx(var h): h.update(data); self = .xx(h)
        case .md5(var h): h.update(data: data); self = .md5(h)
        case .sha1(var h): h.update(data: data); self = .sha1(h)
        case .sha256(var h): h.update(data: data); self = .sha256(h)
        case .sha512(var h): h.update(data: data); self = .sha512(h)
        case .sizeOnly(let n): self = .sizeOnly(n + Int64(data.count))
        }
    }

    /// Gol pentru `.sizeOnly` — la fel ca modelul existent, unde
    /// "verificarea" e doar o comparație de mărime, nu un hash real.
    func finalizeHex() -> String {
        switch self {
        case .xx(let h): return h.hexDigest
        case .md5(let h): return h.finalize().map { String(format: "%02x", $0) }.joined()
        case .sha1(let h): return h.finalize().map { String(format: "%02x", $0) }.joined()
        case .sha256(let h): return h.finalize().map { String(format: "%02x", $0) }.joined()
        case .sha512(let h): return h.finalize().map { String(format: "%02x", $0) }.joined()
        case .sizeOnly: return ""
        }
    }
}

/// Coadă marginită, thread-safe, cu UN singur producător (thread-ul de
/// citire) și UN singur consumator (thread-ul de scriere al destinației
/// respective). `push` blochează dacă adâncimea `capacity` e deja atinsă
/// (backpressure real) — dacă un consumator eșuează, `markConsumerFailed()`
/// oprește orice blocare viitoare, ca thread-ul de citire să nu rămână
/// agățat de o destinație moartă.
final class BoundedChunkQueue {
    private var buffer: [Data] = []
    private let capacity: Int
    private let cond = NSCondition()
    private var finished = false
    private var consumerFailed = false

    init(capacity: Int) { self.capacity = capacity }

    func push(_ chunk: Data) {
        cond.lock()
        defer { cond.unlock() }
        if consumerFailed { return }
        while buffer.count >= capacity && !consumerFailed {
            cond.wait()
        }
        guard !consumerFailed else { return }
        buffer.append(chunk)
        cond.signal()
    }

    func finish() {
        cond.lock(); finished = true; cond.signal(); cond.unlock()
    }

    func markConsumerFailed() {
        cond.lock(); consumerFailed = true; cond.signal(); cond.unlock()
    }

    /// `nil` = EOF normal (sursa s-a terminat, coada s-a golit).
    func pop() -> Data? {
        cond.lock()
        defer { cond.unlock() }
        while buffer.isEmpty && !finished {
            cond.wait()
        }
        guard !buffer.isEmpty else { return nil }
        let chunk = buffer.removeFirst()
        cond.signal() // elibereaza un loc pentru un push care astepta
        return chunk
    }
}

/// [M2, 2026-09-06] Flush FIZIC pe disc — obligatoriu înainte de a marca
/// un fișier "OK". Motiv: apelurile standard de scriere (`FileHandle.write`)
/// scriu doar în cache-ul RAM al sistemului de operare, NU garantează că
/// datele au ajuns pe cipurile flash. Dacă userul scoate cardul/SSD-ul din
/// mufă imediat după ce bara de progres arată 100%, fișierele pot rămâne
/// corupte — exact scenariul pe care un ofloader profesional trebuie să-l
/// elimine matematic, nu doar statistic.
///
/// `fsync(2)` obișnuit NU e suficient pe macOS — documentat oficial de
/// Apple (`man fsync`): pe multe dispozitive de stocare, controllerul
/// hardware raportează scrierea ca „terminată" imediat ce a ajuns în
/// propriul cache electric, ÎNAINTE de a ajunge fizic pe celulele flash.
/// `F_FULLFSYNC` (specific Apple, via `fcntl`) e singurul apel care cere
/// explicit controllerului să golească ACEL cache și să confirme scrierea
/// fizică reală.
///
/// Degradare controlată: unele sisteme de fișiere (volume de rețea SMB/
/// NFS, unele formatări exFAT/FAT32 vechi) NU suportă `F_FULLFSYNC` —
/// întorc `ENOTSUP`. În acel caz, cade pe `fsync()` simplu (tot mai bine
/// decât nimic) — NU tratăm asta ca eroare de transfer. Orice ALTĂ eroare
/// (disc efectiv deconectat, defect) chiar trebuie să oprească fișierul
/// respectiv cu eroare — datele nu sunt confirmate pe disc.
func physicalFlush(_ handle: FileHandle) throws {
    if fcntl(handle.fileDescriptor, F_FULLFSYNC) == -1 {
        let code = errno
        if code == ENOTSUP {
            _ = fsync(handle.fileDescriptor) // fallback - vezi comentariul de mai sus
        } else {
            throw NSError(domain: NSPOSIXErrorDomain, code: Int(code),
                           userInfo: [NSLocalizedDescriptionKey: "Flush fizic eșuat: \(String(cString: strerror(code)))"])
        }
    }
}

final class FanOutCopier {
    private let sourcePath: String
    private let destinationPaths: [String]
    private let chunkSize: Int
    private let ringDepth: Int
    private let model: VerificationModel
    private let cancel: CancelToken
    private let pause: PauseToken

    init(sourcePath: String, destinationPaths: [String], chunkSize: Int,
         ringDepth: Int = 3, model: VerificationModel,
         cancel: CancelToken, pause: PauseToken) {
        self.sourcePath = sourcePath
        self.destinationPaths = destinationPaths
        self.chunkSize = chunkSize
        self.ringDepth = max(2, ringDepth)
        self.model = model
        self.cancel = cancel
        self.pause = pause
    }

    /// `onBytesRead` — apelat de pe thread-ul de citire, o dată per bucată
    /// citită din sursă (pentru progres live în UI, ca înainte).
    func run(onBytesRead: @escaping (Int64) -> Void) throws -> FanOutResult {
        guard let input = FileHandle(forReadingAtPath: sourcePath) else {
            throw FanOutError.cannotOpenSource(sourcePath)
        }
        defer { try? input.close() }

        var queues: [String: BoundedChunkQueue] = [:]
        for dst in destinationPaths {
            queues[dst] = BoundedChunkQueue(capacity: ringDepth)
        }

        var results: [String: FanOutDestOutcome] = [:]
        let resultsLock = NSLock()
        let group = DispatchGroup()

        for dst in destinationPaths {
            group.enter()
            DispatchQueue.global(qos: .userInitiated).async { [self] in
                defer { group.leave() }
                let queue = queues[dst]!
                do {
                    FileManager.default.createFile(atPath: dst, contents: nil)
                    guard let output = FileHandle(forWritingAtPath: dst) else {
                        throw FanOutError.cannotOpenDestination(dst)
                    }
                    defer { try? output.close() }
                    var hasher = IncrementalHasher(model: model)
                    var written: Int64 = 0
                    while let chunk = queue.pop() {
                        try autoreleasepool {
                            try output.write(contentsOf: chunk)
                        }
                        hasher.update(chunk)
                        written += Int64(chunk.count)
                    }
                    // [M2] Flush fizic OBLIGATORIU inainte de a marca fisierul
                    // OK - vezi comentariul din physicalFlush(). O eroare reala
                    // aici (nu ENOTSUP) opreste DOAR aceasta destinatie.
                    try physicalFlush(output)
                    resultsLock.lock()
                    results[dst] = .success(hash: hasher.finalizeHex(), bytesWritten: written)
                    resultsLock.unlock()
                } catch {
                    resultsLock.lock()
                    results[dst] = .failure(error)
                    resultsLock.unlock()
                    queue.markConsumerFailed()
                }
            }
        }

        // Thread-ul de CITIRE — singurul care atinge sursa. Rulează pe
        // thread-ul curent (cel care a apelat `run`), care e deja de
        // fundal in OffloadRunner — nu mai are nevoie de propriul lui
        // DispatchQueue.
        var sourceHasher = IncrementalHasher(model: model)
        var bytesRead: Int64 = 0
        var readError: Error?
        do {
            while true {
                if cancel.isCancelled { throw OffloadCancelled() }
                pause.waitWhilePaused(cancel: cancel)
                if cancel.isCancelled { throw OffloadCancelled() }

                var chunk: Data?
                try autoreleasepool {
                    chunk = try input.read(upToCount: chunkSize)
                }
                guard let chunk, !chunk.isEmpty else { break }

                sourceHasher.update(chunk)
                bytesRead += Int64(chunk.count)
                onBytesRead(Int64(chunk.count))

                for queue in queues.values {
                    queue.push(chunk)
                }
            }
        } catch {
            readError = error
        }
        for queue in queues.values { queue.finish() }

        group.wait()

        if let readError { throw readError }
        return FanOutResult(sourceHash: sourceHasher.finalizeHex(), bytesRead: bytesRead, destinations: results)
    }
}
